### ISSUE TITLE
Linkable buttons

### ISSUE THREAD

**Issue by billthornton** [bbc/gef-docs#75]
Hello!

I *think* this is slightly different than https://github.com/bbc/gef-docs/issues/59 but let me know if it overlaps.

I'm looking at a few of the use-cases where links are being used in place of buttons and trying to see if there is a general improvement that should be applied to them.

### Tabs

One use-case are tabs that link to standalone pages, such as the _date tabs_ on:
* https://www.bbc.co.uk/sport/football/scores-fixtures
* https://www.bbc.co.uk/weather/m1

Some similar implementations offer links instead of buttons if JS is disabled, but that prevents a user from opening one of the tabs in a new browser tab (assuming such a feature is still useful to our users)
* https://www.bbc.co.uk/sport/tennis/results

### More toggle/link

Another use-case is the `More` toggle in the global navigation, which links to the footer page fragment to cater for users without javascript, with the JS enhancing it opening the 'More' drawer. This is read out as a `link` by assistive tech.

### Pagination

Some usage of pagination _may_ also be considered includes in this if - when javascript is enabled - the pagination links change the content in the page rather than linking to a new page
* https://www.bbc.co.uk/sport/live/football/46627053

---

Looking through the current GEF documentation:
* https://bbc.github.io/gef-docs/components/load-more/ seems to contain an approach that hides pagination *links* and displays a show more *button*.
* https://bbc.github.io/gef-docs/components/masthead/#the-more-link allows the more toggle to remain an anchor, whereas the approach in https://bbc.github.io/gef-docs/components/site-menu/ where a button with expanded state is used looks to be more beneficial.
* https://bbc.github.io/gef-docs/components/tabs/ uses same-page links to the tab contents rather than linking to individual pages when JS is not enabled, so I'm not sure if tabs with external links need to be considered here as well?

---

Would you agree that these tabs/toggles should be pronounced as buttons and not as links and function in the same way as buttons do (https://bbc.github.io/gef-docs/components/buttons-and-ctas/#recommended-behaviour)?

There are a few options we can see to achieve this, but non seem ideal.

* Replace the anchor elements as buttons when JS activates (we are using React which makes this easier) which could offer a strange experience if the JS activation takes some time and would prevent the user loading individual pages in a new tab.
* Make a link act like a button using `role=button` and simulating a click even when the space button is pressed - I can guess at this being not widely supported or having some other quirk?

Any thoughts on the above?

**Comment by micmath** [comment]
There is a lot to unpack in the premise of the question, so forgive me if I miss some of the details in my response, but the button v link question is generally a simple one: if it _goes somewhere_ it's a link, if it _does something_ it's a button. There are of course lots of nuances to that in practice: is a form submission "doing something" (submitting data) or "going somewhere" (via a `location` header in the server response, once the data has been processed). But I think the general rule covers 90% of cases.

Another nuance is the single-page application -esque pages, where clicking a thing causes most of the content on the page to change: did the user just _go somewhere_, or was a new page loaded as a result of _doing something_? That's subject to interpretation, and there was a rule used by devs at the BBC that if 80% of the content featured on a page changed, it was considered a new page. Obviously arbitrary, but a rule _is_ needed.

CTA's (the so-called "call to action") pattern is another edge case where the link is presented in a way that is visually similar to a button. This must be based on the assumption that sighted users don't really care what HTML element you used, they just want the result you're pitching to them.

But for screen-reader users I think it does matter, and the decision is not without impact: when to use a button and when to use a link. It would be puzzling to hear _"submit form: link"_ or _"go to about page: button"_.

Often a component is designed to use slow-path actions as a fallback, like an `a` link to a whole new page to display a single new paragraph of text, but the enhancement is to use JavaScript to make an asynchronous request and then update the existing page. In that case the fallback is obviously a link, whereas the enhancement becomes a button: you're no longer _going somewhere_ to see that paragraph.

All of this was considered by @Heydon when the guides were written, but as I showed, some of it is subject to nuance. Are there specific, separate examples where you think we got it wrong?

**Comment by billthornton** [comment]
Thanks @micmath, that's a really good response to my question.

For us the main example was the `More` button in the navigation/masthead (https://bbc.github.io/gef-docs/components/masthead/#the-more-link) where it makes sense for it to be an internal link for non-js, but not the best pattern for a screen reader user with JS. We _could_ switch it out to be a button in javascript, but unsure if this is the best idea.

There's an example of it causing confusion here: https://youtu.be/OUDV1gqs9GA?t=1043

**Comment by Heydon** [comment]
@billthornton @micmath The use of a same-page link here was an effort to make the control usable in the absence of JavaScript (I felt this was important, especially in the case of the main navigation). One other advantage is that, by saying "same-page link" in screen readers they can anticipate a change of focus. However, a similar notion could be communicated with `aria-haspopup="true"` on a button. But then the focus can only be moved with JavaScript.

**Comment by micmath** [comment]
Love the video. @billthornton. Here's a little-known-factoid, I actually was the front end dev that built that damned More button in Barlesque (as it was known at the time). So I feel personally invested in this discussion suddenly :-)

But that is an error that proves the rule actually. The way the More menu was designed to work is that it really was an in-page link, that properly just moved focus to another part of the page, a list of further options located just above the footer. But in the JE-enabled _enhanced_ version of the page we wrote a script to move that list of links from the bottom of the page up into a hidden drawer. And the enhanced experience for the JS-blessed was that they go a nice list of navigation links, and a slidey drawer effect that held the overflow links.

So under that design the un-enhanced thing really is a link (although and in-page one). Where we went wrong was in not upgrading it to a button for the enhanced experience. And the proof that this was a Bad Thing is in your video. Screen reader users actually know when you get the semantics wrong, and it actually can cause confusion.

**Comment by billthornton** [comment]
I definitely understand (and agree with) the non-JS behaviour! As noted in the original post, it's easy for us to `Replace the anchor elements as buttons when JS activates`. Would that seem like a reasonable solution to provide the best experience for both non-js and js?

**Comment by micmath** [comment]
I'm hesitant to answer without actual data, but the video makes pretty strong case. I'd say that if an element walks and talks like a button when JS is active then it should be a button.

**Comment by Heydon** [comment]
@billthornton @micmath I don't think there is a behaviour/semantics mismatch since the same-page link works as expected. Whether a button would be more appropriate/expected in any case is arguable, though. I have a hunch that users are generally more familiar with same-page links than `aria-haspopup` buttons, since the first is a much longer established paradigm. 

If we switched from same-page link to `aria-haspopup` button, we'd essentially be switching from two different, but equally technically sound, solutions for the same focus management issue. I'm leaning towards not bothering, since it increases complexity in the component.

**Comment by micmath** [comment]
I admit, I don't fully understand the complexities of all the arguments being made here, but @Heydon are you saying that because, when the element is clicked, focus moves to another part of the page, then it is semantically equivalent to an in-page link, and _therefore_ using an A ELEMENT is valid?

The fact that the other part of the page is hidden, and slides open to be visible when it _receives_ focus is a red herring in that case, and the _if it does something_ rule doesn't strictly apply?

Sorry, trying to wrap my brain around this!

**Comment by Heydon** [comment]
Yep, that's what I'm thinking. But you raise a good point about the disclosure behavior (I was still picturing the "everything open" no JS mode for some reason!)

That considered, I think the most robust and least intensive enhancement would be to add role=button, aria-haspopup=true, and the space bar activation support.
